mocks/mocks.py

```diff
@@ -18,15 +18,172 @@
+    config = utils.load_yaml_config(config_file)
+    # Coordinates of the target galaxy.
+    config['galaxy_coordinates'] = kwargs.get('galaxy_coordinates',
+                                              config['galaxy_coordinates'])
+    # The time of the observation.
+    config['observation_time'] = kwargs.get('observation_time',
+                                            config['observation_time'])
+    # The path to the data.
+    config['data_path'] = kwargs.get('data_path',
+                                     config['data_path'])
+    config['imager_filter'] = kwargs.get('imager_filter',
+                                         config['imager_filter'])
+    config['mass_to_light_ratio'] = kwargs.get('mass_to_light_ratio',
+                                               config['mass_to_light_ratio'])
+    config['abs_mag_sun'] = kwargs.get('abs_mag_sun',
+                                       config['abs_mag_sun'])
+    # The distance between the observer and the target galaxy. This is
+    # luminosity distance.
+    config['galaxy_distance'] = kwargs.get('galaxy_distance',
+                                           config['galaxy_distance'])
+    config['galaxy_distance'] = ensure_unit(config['galaxy_distance'], u.Mpc)
+    # The position of the target galaxy along the projection direction in the
+    # simulation box. Since it is coming from simulations, it is co-moving
+    # distance in Mpc.
+    config['target_galaxy_comoving_depth'] =\
+        kwargs.get('target_galaxy_comoving_depth',
+                   config['target_galaxy_comoving_depth'])
+    config['target_galaxy_comoving_depth'] = ensure_unit(
+        config['target_galaxy_comoving_depth'], u.Mpc)
+    config['viewing_axis'] = kwargs.get('viewing_axis',
+                                        config['viewing_axis'])
+    # The resolution of the simulation(parsec / pixel).
+    config['sim_pc_pixel'] = kwargs.get('sim_pc_pixel',
+                                        config['sim_pc_pixel'])
+    config['sim_pc_pixel'] = ensure_unit(config['sim_pc_pixel'],
+                                         u.pc / u.pixel)
+    config['particle_baryonic_mass_sim'] =\
+        kwargs.get('particle_baryonic_mass_sim',
+                   config['particle_baryonic_mass_sim'])
+    # To Do: The Hubble constant in the simulations is h, which is H / 100, it
+    # should be checked again that what constant is used in where and make
+    # proper conversion where it applies.
+    try:
+        config['hubble_constant'] = kwargs.get('hubble_constant',
+                                               config['hubble_constant'])
+    except KeyError:
+        config['hubble_constant'] = default_cosmology.H0
+    try:
+        assert config['hubble_constant']
+    except AssertionError:
+        config['hubble_constant'] = default_cosmology.H0
+    config['hubble_constant'] = ensure_unit(config['hubble_constant'],
+                                            u.km / (u.Mpc * u.s))
```

Comment:
For cosmological parameters, the code first tries to get the value from the `kwargs` and if it does not exist, the code will try the `config`. If the key does not exist in the `config`, it will except the `KeyError` and set the value by default cosmology.
The second `try .... expext .... ` works when the parameter is defined in the configuration file but has no value. This block will set the value using the default cosmology and prevent the existence of cosmological parameters with value `None`.
I have not done the same for keys other than the cosmological parameters because I thought an error should raise when they have the value `None`, but I am not sure if the error should be raised here in the `parse_config`.

Reply:
See comment above about a neater way to do this. In this case it would become:
```
config['hubble_constant'] = kwargs.get('hubble_constant',
                                        config.get('hubble_constant',
                                                    default_cosmology.H0))
```

I don't think you need to worry about the case where the key is in the config file but has no value. I think it's fair to just consider that a bad config file.  If the user wants to revert to using default values without removing the key from the config file they should just comment the line, not remove the value.

Reply:
@bazkiaei I don't think we need to worry about the case where they don't provide part of the cosmological parameters, since the mock code is no longer doing what the user wants - converting a simulation into something realistic. SImilarly, there is no case where we will send in cosmological data that should default to a cosmology that you've selected here - *all* cosmological simulations will have been run with a cosmology. If that cosmology is unknown, then the simulation is worthless to us. If that cosmology is known, then we can just require the user to provide the cosmology exactly.

To summarise, there is no use case to build in a flexible or default cosmology as the cosmological data will always have a cosmology that is known perfectly.